### PR TITLE
refactor gradient API

### PR DESCRIPTION
**Pull request by toaster**

### CODE REVIEW

File: canvas/gradient.go [hunk 1]
```diff
@@ -62,99 +114,40 @@ func (g *Gradient) calculatePixel(w, h, x, y float64) *color.RGBA64 {
...
-type pixelGradient struct {
-	g   *Gradient
-	img draw.Image // image cache during generator run
 }
 
-// NewLinearGradient returns a new Image instance that dynamically
-// renders a gradient of type GradientDirection
-func NewLinearGradient(start color.Color, end color.Color, direction GradientDirection) *Gradient {
-	pix := &pixelGradient{}
+func computeGradient(g Gradient, w, h int, startColor, endColor color.Color) image.Image {
```
Comment (andydotxyz) [inline canvas/gradient.go]: I think that here the startColor and endColor are superfluous as it's set in all Gradient types.
If Gradient is still an interface you could pass baseGradient, or if it's changed as above then the colours can be read directly

File: canvas/gradient.go [hunk 2]
```diff
@@ -3,49 +3,101 @@ package canvas
...
-	GradientDirectionHorizontal
-	// GradientDirectionCircular defines a gradient traveling radially from a center point outward
-	GradientDirectionCircular
+	"fyne.io/fyne"
 )
 
-// GradientDirection defines a starting or stopping location
-type GradientDirection int
+// Gradient describes a gradient fade between two colors.
+type Gradient interface {
```
Comment (andydotxyz) [inline canvas/gradient.go]: I'm not sure about the interface here - CanvasObjects are structs due to their concrete nature.
I see that users can create &LinearGradient{} or &RadialGradient{} so maybe it's not an issue - did you consider this in the design?

File: canvas/gradient.go [hunk 3]
```diff
@@ -3,49 +3,101 @@ package canvas
...
+	fyne.CanvasObject
+	calculateGradient(x, y, w, h float64) float64
+	Compute(w, h int) image.Image
```
Comment (andydotxyz) [inline canvas/gradient.go]: This moved from Generator before and the comment about not using directly - has the intent of the method changed?
As I understood it this could be moving into the driver for optimisation purposes and so people were warned in the docs. Do you think this has changed now?

File: canvas/gradient_test.go [hunk 4]
```diff
@@ -7,35 +7,30 @@ import (
...
-func TestNewLinearGradient(t *testing.T) {
-	// Horizontal
-	horizontal := NewLinearGradient(color.Black, color.Transparent, GradientDirectionHorizontal)
-	assert.Equal(t, horizontal.CenterOffsetX, 0.0)
-	assert.Equal(t, horizontal.CenterOffsetY, 0.0)
+func TestNewHorizontalGradient(t *testing.T) {
+	horizontal := NewHorizontalGradient(color.Black, color.Transparent)
 
-	img := horizontal.Generator(50, 5)
+	img := horizontal.Compute(50, 5)
```
Comment (andydotxyz) [inline canvas/gradient_test.go]: As above is this change required? The new name sounds more callable than the previous Generator... (and is a new term compared to the Raster.Generator).

File: canvas/gradient.go (hunk 3 above)
Comment (toaster) [reply canvas/gradient.go]: `Generator` returns a `func` which is okay for `canvas.Raster` which does not know anything about its generation (the function is handed in on creation of the `Raster`). But for the gradients I can't see any advantage in the indirection of returning a generator method instead of just perform the computation.

**Comment by toaster** [comment]
I removed the `Gradient` interface and simplified the gradient structs.
